### PR TITLE
Switch to flake

### PR DESCRIPTION
**Pull request by L-as**
This replaces all the existing Nix code with a flake. I haven't tested this thoroughly yet.

You need at least Nix 2.4 for this to work.

`nix develop` *should* open a shell.
`nix flake check` *should* run the tests.

Updating dependencies:
I will probably make a guide for this, but in general you should choose a commit of **plutus-apps**,
use that in `cabal.project`, then make sure that the other git repositories you depend on use the same commit as specified in the `cabal.project` of `plutus-apps` in the commit you've chosen. This is to ensure that we use the same versions of the dependencies as upstream expects.
After this, haskell.nix will complain about missing entries in the `sha256map` in `flake.nix`. Fixing this is easy but tedious, you must make a new entry for the commit you're using, remove the old one, and initially set it to a "random" hash, e.g. 52 0s.
Nix will tell you the hash it expected, you can paste that in in place of the hash you specified.
It is important that the "random" hash you choose is not generated by hashing something, such that Nix can not fetch it from a cache.

### CODE REVIEW

**Comment by L-as** [comment]
Also, I've enabled `exactDeps = true` here. This should mean you don't have to build the dependencies in your shell, but it didn't work last time I tried it, so it might need some adjustments.

**Comment by L-as** [comment]
`exactDeps` doesn't work after all, which is a huge shame. It complains about missing dependencies of `plutus-doc`, which makes absolutely no sense. I will look into this in the future. `nix flake check` doesn't work because it tries to evaluate the derivations for other systems, which doesn't work since haskell.nix uses IFD.

**Comment by L-as** [comment]
You can run the tests by doing `nix build .#check.x86_64-linux`. `nix flake show` can not work due to issues with Nix, IFD, and haskell.nix. `nix develop` seems to work fine. I will not try to fix this issue nor the `exactDeps` ones until we get confirmation that upstream are willing to support us and include our fixes.

File: flake.nix [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+  description = "plutus-extra";
+
+  inputs.haskell-nix.url = "github:input-output-hk/haskell.nix";
+  inputs.nixpkgs.follows = "haskell-nix/nixpkgs-unstable";
+  inputs.haskell-nix.inputs.nixpkgs.follows = "haskell-nix/nixpkgs-2105";
+  inputs.plutus.url = "github:input-output-hk/plutus"; # used for libsodium-vrf
+
+  outputs = { self, nixpkgs, haskell-nix, plutus }:
+    let
+      supportedSystems = [ "x86_64-linux" "x86_64-darwin" "aarch64-linux" "aarch64-darwin" ];
```
Comment (kozross) [inline flake.nix:11]: Does Plutus actually support aarch64?

**Review by kozross — APPROVED** [review]

**Pull request merged**